src/transformers/utils/model_parallel_utils.py

```diff
@@ -12,10 +12,176 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
+import copy
+import importlib
 from math import ceil
 
 
+class TPInfo(object):
+    def __init__(
+        self,
+        *name,
+        combined_qkv: bool = False,
+        reverse: bool = False,
+    ):
+        self.name = name
+        self.combined_qkv = combined_qkv
+        self.reverse = reverse
+
+    def __str__(self):
+        return f"{self.__class__.__qualname__}({self.name})"
+
+    def __repr__(self):
+        return self.__str__()
+
+
+Col = type("COLUMN", (TPInfo,), {"code": "Col"})
+Row = type("ROW", (TPInfo,), {"code": "Row"})
+Update = type("UPDATE", (TPInfo,), {"code": "Update"})
+
+
+class TPMapping(object):
+    __MAPPING__ = dict(
+        Albert=[
+            Col("query", "key", "value", "ffn"),
+            Row("dense", "ffn_output"),
+            Update("num_attention_heads", "all_head_size"),
+        ],
+        Bart=[
+            Col("q_proj", "k_proj", "v_proj", "fc1"),
+            Row("out_proj", "fc2"),
+            Update("embed_dim", "num_heads"),
+        ],
+        Bert=[
+            Col("query", "key", "value", "intermediate.dense"),
+            Row("output.dense"),
+            Update("num_attention_heads", "all_head_size"),
+        ],
+        T5=[
+            Col("q", "k", "v", "DenseReluDense.wi"),
+            Row("o", "DenseReluDense.wo", "relative_attention_bias"),
+            Update("d_model", "n_heads", "inner_dim"),
+        ],
+        GPT2=[
+            Col("c_attn", reverse=True, combined_qkv=True),
+            Col("q_attn", reverse=True),
+            Row("c_proj", reverse=True),
+            Update("embed_dim", "split_size", "num_heads"),
+        ],
+        Electra=[
+            Col("query", "key", "value", "intermediate.dense"),
+            Row("output.dense"),
+            Update("num_attention_heads", "all_head_size"),
+        ],
+        Roberta=[
+            Col("query", "key", "value", "intermediate.dense"),
+            Row("output.dense"),
+            Update("num_attention_heads", "all_head_size"),
+        ],
+    )
+
+    def __init__(self):
+        cache_tp_mapping = {}
+
+        for cls_name, mapping in self.__MAPPING__.items():
+            cls = self._load_class_by_model_name(cls_name)
+            cache_tp_mapping[cls] = []
+
+            for elem in mapping:
+                for name in elem.name:
+                    copy_elem = copy.deepcopy(elem)
+                    copy_elem.name = name
+                    cache_tp_mapping[cls].append(copy_elem)
+
+        self.__MAPPING__ = {cls: {} for cls in cache_tp_mapping}
+        # clear exist mapping rather than making new mapping dict
+
+        for cls, mapping in cache_tp_mapping.items():
+            for elem in mapping:
+                if elem.code in self.__MAPPING__[cls]:
+                    self.__MAPPING__[cls][elem.code].append(elem)
+                else:
```

Comment:
How about using a `defaultdict` in line 96, like 

```python
# from collections import defaultdict

self.__MAPPING__ = {cls: defaultdict(list) for cls in cache_tp_mapping}
```

so that we can remove the `else` here?

Reply:
Looks good to me !